### ISSUE TITLE
Third party library updates

### ISSUE THREAD

**Issue by sgpearse** [NCAR/VAPOR#3307]
Third party libraries need to be updated to their latest versions.  Specific versions TBD.

**Comment by shaomeng** [comment]
We should look into spack (https://spack.io/) because this is the tool for building packages from source. It could potentially reduce our workload significantlyl

**Comment by sgpearse** [comment]
@shaomeng and @NihanthCW - I've been trying spack and I'm hitting snags with building hdf5 on both my OSX and Ubuntu20 development machines.  Installing hdf5 is [their bare-bones example](https://spack.io/about/), which is not encouraging.

At this point, I've sunk a non-trivial amount of time into rebuilding the third party libraries in the same way that has been done in the past.  I would personally like to continue with our protocol rather than interrupt it with a new research experiment.  Our code freeze is quickly approaching and IMO, this is a time bad for experimentation.

Sam, since you're more versed with spack and would like to research new options, I can put this issue aside and await your findings.  There are other issues that will be more productive for me to work on if we insist that this needs to be researched right now.  I will put this work on hold until we have a consensus.

**Comment by shaomeng** [comment]
Sounds good @sgpearse , I can look into if spack is a good fit for us!

**Comment by shaomeng** [comment]
Also @sgpearse , how's your circle ci approach going? You mentioned that you're half way there on the circle ci path, right?

**Comment by sgpearse** [comment]
> You mentioned that you're half way there on the circle ci path, right?

No, I didn't say that.  I've embarked on rebuilding them in the same way we've done in the past with some improvements.

In today's meeting I said that I would like for the libraries to be generated on CI systems for reproducibility reasons.

If it's helpful, the Google Drive for our libraries has [a spreadsheet for the versions we're targetting for 3.8.1](https://docs.google.com/spreadsheets/d/16caPgiDOIfrzEo2UGH7HD-N8O43y4iuwkri9noYEzE8/edit#gid=0).

**Comment by NihanthCW** [comment]
> @shaomeng and @NihanthCW - I've been trying spack and I'm hitting snags with building hdf5 on both my OSX and Ubuntu20 development machines. Installing hdf5 is [their bare-bones example](https://spack.io/about/), which is not encouraging.
> 
> At this point, I've sunk a non-trivial amount of time into rebuilding the third party libraries in the same way that has been done in the past. I would personally like to continue with our protocol rather than interrupt it with a new research experiment. Our code freeze is quickly approaching and IMO, this is a time bad for experimentation.
> 
> Sam, since you're more versed with spack and would like to research new options, I can put this issue aside and await your findings. There are other issues that will be more productive for me to work on if we insist that this needs to be researched right now. I will put this work on hold until we have a consensus.

Hi Scott, First of all, I appreciate you putting the effort to see if spack works for us. Just to clarify: We discussed this in the meeting and decided NOT to experiment with or take this decision before this release, right? I am curious to know what led you to believe that you should implement this now.  

I agree that any new tool must be experimented with prior to adoption and that was indeed what was decided especially with spack's lack of Windows support from what I remember. We weren't even sure if this is the right tool. I think Sam's earlier comment here was from before the meeting. We can have a discussion on how/who needs to work on this at a later stage. Thanks!

**Comment by sgpearse** [comment]
> Hi Scott, First of all, I appreciate you putting the effort to see if spack works for us. Just to clarify: We discussed this in the meeting and decided NOT to experiment with or take this decision before this release, right? I am curious to know what led you to believe that you should implement this now.

Thanks for your appreciation Nihanth.  My recollection is that we deliberated using spack and then agreed to explore it.  Sam added a comment to this issue **_during our meeting_**, which indicated to me that this is an addendum to the current requirements.  That's why I pursued it.

We've touched on this this kind of workflow conundrum recently via email.  IMO, spack should have been written up as an entirely new issue and not a supplement to the current one.

**Comment by shaomeng** [comment]
Hey @sgpearse sorry if I caused confusion to you, but what I intended to do and still think what I did, both in my GitHub comment and in Google Meeting comments, was to suggest an alternate approach rather than to mandate a change of course. If you could let me know what exactly made you feel compelled to change course, I'll certainly improve on my future communication.

**Comment by NihanthCW** [comment]
@sgpearse I feel I am missing something in this particular situation. I am still not clear on how Sam's comment could be interpreted as a mandate. Even more so in light of the prolonged discussion that we had in the meeting today about spack and agreeing to explore it after the release. I think this is can be better resolved in a meeting rather than here. Let's discuss. 

Nonetheless, I am glad you reached out for clarification sooner than later. Looks like we have a path forward - i.e., Continue with your original plan (how you've always updated third party libraries). 

I understand that updating third party libraries is a pain point that you've brought up numerous times before. I encourage and welcome suggestions on how we can improve the process at anytime. However, the decision on how to proceed will only be made after mutual agreement in the team and following the process that we have in place. Thanks!

**Comment by sgpearse** [comment]
This is just a miscommunication.  I'll keep moving forward with the standard build routine.

Exploring spack seems like a mandate because the github comment says "we should do X" on a ticket with a 3.8.1 milestone.  The comment has even more weight because it was made during our meeting, and comments with this kind of timing typically occur when the team reaches a consensus on something.  I also remember saying "I'll look into it" as the closing comment in our discussion on this yesterday.

However I don't remember us specifically stating whether we would explore this before or after release, but the fact that this is a ticket with a 3.8.1 milestone implies that we do it before.  Either way, I still think it was worth trying.  If we can build all of the libraries for a given platform with 18 commands, that's a big win even if we can't use it for Windows.

I apologize if I missed the agreement that this was a post-3.8.1 action item, but if that's the case, we should create a new ticket for spack.
